tracing-modality/src/lib.rs

```diff
@@ -35,32 +36,38 @@ pub enum InitError {
 
 /// A global tracer instance for [tracing.rs](https://tracing.rs/) that sends traces via a network
 /// socket to [Modality](https://auxon.io/).
-pub struct TracingModality {}
+pub struct TracingModality {
+    ingest_handle: ModalityIngestHandle,
+}
 
 impl TracingModality {
     /// Initialize with default options and set as the global default tracer.
     pub fn init() -> Result<Self, InitError> {
-        let disp = Dispatch::new(TSSubscriber::new());
-        tracing::dispatcher::set_global_default(disp).unwrap();
-
-        TSSubscriber::connect().context("connect to modality")?;
-
-        Ok(Self {})
+        Self::init_with_options(Default::default())
     }
 
     /// Initialize with the provided options and set as the global default tracer.
-    pub fn init_with_options(opt: Options) -> Result<Self, InitError> {
-        let disp = Dispatch::new(TSSubscriber::new_with_options(opt));
+    pub fn init_with_options(opts: Options) -> Result<Self, InitError> {
+        let mut layer =
+            ModalityLayer::init_with_options(opts).context("initialize ModalityLayer")?;
+        let ingest_handle = layer
+            .take_handle()
+            .expect("take handle on brand new layer somehow failed");
+
+        let disp = Dispatch::new(layer.into_subscriber());
         tracing::dispatcher::set_global_default(disp).unwrap();
 
-        TSSubscriber::connect().context("connect to modality")?;
+        Ok(Self { ingest_handle })
+    }
 
-        Ok(Self {})
+    /// Stop accepting new trace events, flush all existing events, and stop ingest thread.
+    pub fn finish(&mut self) {
```

Comment:
If this should only ever be called once, maybe this method should accept `self` rather than `&mut self`